syntax = "proto2";

enum ServerType {
  GOSSIP = 0;
  SEGMENTED = 1;
  PAXOS = 2;
}

// Server.
message BenchmarkServerStartRequest {
  enum Object {
    TWO_INTS = 0;
    BANK_ACCOUNT = 1;
  }

  required uint64 num_servers = 1;
  required Object object = 2;
  required ServerType server_type = 3;
}

message BenchmarkServerStartReply {
  required uint64 index = 1;
}

message BenchmarkServerKillRequest {
}

message BenchmarkServerKillReply {
  required uint64 index = 1;
}

message BenchmarkServerRequest {
  oneof request {
    BenchmarkServerStartRequest start = 1;
    BenchmarkServerKillRequest kill = 2;
  }
}

message BenchmarkServerReply {
  oneof reply {
    BenchmarkServerStartReply start = 1;
    BenchmarkServerKillReply kill = 2;
  }
}

// Client.
message BenchmarkClientBankAccountRequest {
  required uint64 num_servers = 1;
  required double fraction_withdraw = 2;
  required uint64 duration_in_milliseconds = 3;
  required ServerType server_type = 4;
}

message BenchmarkClientBankAccountReply {
  required uint64 index = 1;
  required uint64 num_txns = 2;
  required uint64 duration_in_nanoseconds = 3;
  required double txns_per_second = 4;
}

message BenchmarkClientTwoIntsRequest {
  required uint64 num_servers = 1;
  required uint64 duration_in_milliseconds = 2;
  required ServerType server_type = 3;
}

message BenchmarkClientTwoIntsReply {
  required uint64 index = 1;
  required uint64 num_txns = 2;
  required uint64 duration_in_nanoseconds = 3;
  required double txns_per_second = 4;
}

message BenchmarkClientRequest {
  oneof request {
    BenchmarkClientBankAccountRequest bank_account = 1;
    BenchmarkClientTwoIntsRequest two_ints = 2;
  }
}

message BenchmarkClientReply {
  oneof reply {
    BenchmarkClientBankAccountReply bank_account = 1;
    BenchmarkClientTwoIntsReply two_ints = 2;
  }
}
